src/Hash.cc

```diff
@@ -206,16 +197,273 @@ void* HashKey::TakeKey()
 		return CopyKey(key, size);
 	}
 
-void* HashKey::CopyKey(const void* k, int s) const
+char* HashKey::CopyKey(const char* k, size_t s) const
 	{
-	void* k_copy = (void*) new char[s];
+	char* k_copy = new char[s];
 	memcpy(k_copy, k, s);
 	return k_copy;
 	}
 
-hash_t HashKey::HashBytes(const void* bytes, int size)
+hash_t HashKey::HashBytes(const void* bytes, size_t size)
 	{
 	return KeyedHash::Hash64(bytes, size);
 	}
 
+void HashKey::Set(bool b)
+	{
+	Set(b);
+	}
+
+void HashKey::Set(int i)
+	{
+	key_u.i = i;
+	key = reinterpret_cast<char*>(&key_u);
+	size = sizeof(i);
+	}
+
+void HashKey::Set(bro_int_t bi)
+	{
+	key_u.bi = bi;
+	key = reinterpret_cast<char*>(&key_u);
+	size = sizeof(bi);
+	}
+
+void HashKey::Set(bro_uint_t bu)
+	{
+	key_u.bi = bro_int_t(bu);
+	key = reinterpret_cast<char*>(&key_u);
+	size = sizeof(bu);
+	}
+
+void HashKey::Set(uint32_t u)
+	{
+	key_u.u32 = u;
+	key = reinterpret_cast<char*>(&key_u);
+	size = sizeof(u);
+	}
+
+void HashKey::Set(double d)
+	{
+	union {
+		double d;
+		int i[2];
+	} u;
+
+	key_u.d = u.d = d;
+	key = reinterpret_cast<char*>(&key_u);
+	size = sizeof(d);
+	}
+
+void HashKey::Set(const void* p)
+	{
+	key_u.p = p;
+	key = reinterpret_cast<char*>(&key_u);
+	size = sizeof(p);
+	}
+
+void HashKey::Reserve(const char* tag, size_t addl_size, size_t alignment)
+	{
+	ASSERT(! IsAllocated());
+	size_t s0 = size;
+	size_t s1 = util::memory_size_align(size, alignment);
+	size = s1 + addl_size;
+	}
+
+void HashKey::Allocate()
+	{
+	if ( key != nullptr and key != reinterpret_cast<char*>(&key_u) )
+		{
+		reporter->InternalWarning("usage error in HashKey::Allocate(): already allocated");
+		return;
+		}
+
+	if ( size > 0 )
+		{
+		is_our_dynamic = true;
+		key = reinterpret_cast<char*>(new double[size/sizeof(double) + 1]);
+		}
+
+	read_size = 0;
+	write_size = 0;
+	}
+
+void HashKey::Write(const char* tag, bool b)
+	{
+	Write(tag, &b, sizeof(b), 0);
+	}
+
+void HashKey::Write(const char* tag, int i, bool align)
+	{
+	if ( ! IsAllocated() )
+		{
+		Set(i);
+		return;
+		}
+
+	Write(tag, &i, sizeof(i), align ? sizeof(i) : 0);
+	}
+
+void HashKey::Write(const char* tag, bro_int_t bi, bool align)
+	{
+	if ( ! IsAllocated() )
+		{
+		Set(bi);
+		return;
+		}
+
+	Write(tag, &bi, sizeof(bi), align ? sizeof(bi) : 0);
+	}
+
+void HashKey::Write(const char* tag, bro_uint_t bu, bool align)
+	{
+	if ( ! IsAllocated() )
+		{
+		Set(bu);
+		return;
+		}
+
+	Write(tag, &bu, sizeof(bu), align ? sizeof(bu) : 0);
+	}
+
+void HashKey::Write(const char* tag, uint32_t u, bool align)
+	{
+	if ( ! IsAllocated() )
+		{
+		Set(u);
+		return;
+		}
+
+	Write(tag, &u, sizeof(u), align ? sizeof(u) : 0);
+	}
+
+void HashKey::Write(const char* tag, double d, bool align)
+	{
+	if ( ! IsAllocated() )
+		{
+		Set(d);
+		return;
+		}
+
+	Write(tag, &d, sizeof(d), align ? sizeof(d) : 0);
+	}
+
+void HashKey::Write(const char* tag, const void *bytes, size_t n, size_t alignment)
```

Comment:
Is `tag` intentionally unused here? Same question for `SkipWrite`, `Read`, and `SkipRead` below. This has follow-on effects for the other `Write`, etc methods.

Reply:
Quick recap here for anyone reading since Tim and I discussed in person: this mirrors the use of tagging as a troubleshooting aid in `SerializationFormat`. The tags (short, static strings) are only used in debug-stream log output, but remain in production builds.

Reply:
I will clarify that in a comment.